npep/npep-126-egress-traffic-control.md

```diff
@@ -0,0 +1,100 @@
+# NPEP-126: Add northbound traffic support in (B)ANP API
+
+* Issue: [#126](https://github.com/kubernetes-sigs/network-policy-api/issues/126)
+* Status: Provisional
+
+## TLDR
+
+This NPEP proposes adding support for cluster egress (northbound) traffic control in the `AdminNetworkPolicy` and `BaselineAdminNetworkPolicy` API objects.
+
+## Goals
+
+* Implement egress traffic control towards external destinations (outside the cluster)
+* Implement egress traffic control towards cluster nodes
+  - Currently the behaviour for policies defined around traffic from cluster
+    workloads (non-hostNetworked pods) towards nodes in the
+    cluster is undefined. See https://github.com/kubernetes-sigs/network-policy-api/issues/73.
+    - ANP currently supports only east-west traffic and this traffic flow cuts from
+    overlay to underlay which makes this part of the egress (northbound) use case.
+    - Let's provide a defined behaviour in ANP to explicitly achieve the use case.
+    - NOTE1: Traffic towards nodes here includes traffic towards host-networked pods on that node
+      because a "node" resource encompasses all objects that share the host-networking resources
+    - NOTE2: Traffic towards nodes here exluding traffic towards host-networked pods is outside
+      the scope of this NPEP and is a non-goal
+* Implement egress traffic control towards hostNetworked pods
+  - Currently the behaviour for policies defined around traffic from cluster
+  workloads (non-hostNetworked pods) towards hostNetworked pods in the
+  cluster is undefined. See https://github.com/kubernetes-sigs/network-policy-api/issues/73.
+  - ANP currently supports only east-west traffic and this traffic flow cuts from
+  overlay to underlay which makes this part of the egress (northbound) use case.
+  - Let's provide a defined behaviour in ANP to explicitly achieve the use case.
+  - NOTE: We can keep this goal experimental in case implementors cannot distinguish between
+    a node and host-networked pod
+
+## Non-Goals
+
+* Implementing southbound traffic use cases is outside the scope of this NPEP
+* Implementing specific api-server user stories where the definition of an api-server
+matches neither a nodeSelector not a podSelector or any expressive external destinaton
+is outside the scope of thie NPEP and perhaps warrants a new NPEP.
+* Implementing traffic towards nodes exluding traffic towards host-networked pods on that
+node is outside the scope of this NPEP and is a non-goal
+
+## Introduction
+
+### User Stories for egress traffic control towards external destinations
```

Comment:
I think FQDN selectors for egress are a important use-case to include here.

As a cluster administrator I want to restrict traffic to specific destinations outside the cluster using a fully-qualified domain name.
Example: Pods in namespaceA is allowed to send traffic to *.github.com.

Reply:
To close the loop, we decided to spin this discussion out into its own NPEP.

https://github.com/kubernetes-sigs/network-policy-api/issues/133

Reply:
thanks @rahulkjoshi !! :)